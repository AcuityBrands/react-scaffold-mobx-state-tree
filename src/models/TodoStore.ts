import { types, getParent, destroy, flow, getSnapshot } from "mobx-state-tree"
import { createHistory } from './History'

export type ITodoItem = typeof TodoItem.Type;
export type ITodoStore = typeof TodoStore.Type;

// A TodoItem contains the information describing a Task
export const TodoItem = types
  .model("Todo", {
    userId: types.optional(types.number, 1),
    id: types.number,
    title: types.string,
    completed: types.optional(types.boolean, false)
  })
  .actions(self => ({
    // Change title given a new title
    changeLabel(newLabel: string) {
      self.title = newLabel;
    },
    // Change the completed value given a new completed value
    changeComplete(newValue: boolean) {
      self.completed = newValue;
    },
    // Navigate up dependence tree two levels and remove self from tree
    remove() {
      // Level 1 = array of todos
      // Level 2 = todo store
      getParent(self, 2).remove(self);
    },
    // Use mobx generator function to create synchronous async action
    // Similiar to async / await
    // This method mocked for reference only - never used in the app
    save: flow(function* save() {
      try {
        yield window.fetch(`https://jsonplaceholder.typicode.com/todos/1`, {
          method: "PUT",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(getSnapshot(self))
        })
      } catch (ex) {
        console.warn("Unable to save todo", ex)
      }
    })
  }))

// A TodoStore's primary responsibility is to store an array of TodoItems
// This model is composed of the typical stuff (models, actions, views)
// But also a generic history class that allows undo and redo actions
export const TodoStore = types.compose(types
  .model("TodoStore", {
    todos: types.optional(types.array(TodoItem), []),
    loading: types.optional(types.boolean, false)
  })
  .actions(self => ({
    // Add todo to todo array
    add(todo: ITodoItem) {
      self.todos.push(todo);
    },
    // Remove todo from todo array
    remove(todo: ITodoItem) {
      destroy(todo);
    },
    // Load contrived todo suggestions into todo array
    // Adding randomized id to avoid React 'key' errors
    getSuggestions: flow(function* getSuggestions(count: number) {
      self.loading = true;
      try {
        const response = yield window.fetch('data/todos.json');
        const todos = yield response.json();

        // Grab requested number since service returns 200 records
        const slice = todos.slice(0, count); 

        // Randomize the ID
        slice.map((item: ITodoItem) => { 
          item.id = Math.floor(Math.random() * 100000000)
        })

        self.todos.push(...slice);
        self.loading = false;
      } catch (ex) {
        console.error("Error while loading suggestions:", ex);
        self.loading = false;
      }
    })
  }))
  .views(self => {
    // View functions wrapped in literal to allow referencing 
    // internal properties from within other functions
    // Only works if accessed property is a getter using 'get'
    // Other options found here: https://github.com/mobxjs/mobx-state-tree#lifecycle-hooks-for-typesmodel
    const vx = {
      get todoCount() {
        return self.todos.length;
      },
      get completedCount() {
        return self.todos.filter(todo => todo.completed == true).length;
      },
      get completedPct() {
        if (vx.todoCount == 0) return 0;
        return Math.floor((vx.completedCount / vx.todoCount) * 100);
      }
    }
    return vx;
  }),
  createHistory(["/loading"])
)