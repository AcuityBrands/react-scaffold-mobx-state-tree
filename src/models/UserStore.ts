import { types, getParent, destroy, flow, getSnapshot } from "mobx-state-tree"

export type IUserStore = typeof UserStore.Type;

export const User = types
  .model("User", {
    id: types.optional(types.number, -1),
    name: types.optional(types.string, '')
  })

export const UserStore = types
  .model("UserStore", {
    user: types.optional(User,{}),
    isAuthenticated: types.optional(types.boolean, false),
    authenticating: types.optional(types.boolean, false)
  })
  .actions(self => ({
    // Load contrived sales data into store
    login: flow(function* authenticateUser(username?: string, password?: string) {
      self.authenticating = true;
      try {
        const response = yield window.fetch('data/user.json');
        const data = yield response.json();
        self.user = User.create(data);
        self.authenticating = false;
        self.isAuthenticated = true;
      } catch (ex) {
        console.error("Error while loading user data:", ex);
        self.authenticating = false;
      }
    }),
    logout: () => {
      self.isAuthenticated = false;
    }
  }))
