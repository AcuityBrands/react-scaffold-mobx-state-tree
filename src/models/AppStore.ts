import { types, getEnv } from "mobx-state-tree"
import { ChartStore } from "./ChartStore"
import { TodoStore } from "./TodoStore"

export type IAppStore = typeof AppStore.Type

// AppStore is a store of stores.
export const AppStore = types
  .model("AppStore", {
    chartStore: types.optional(ChartStore, {}),
    todoStore: types.optional(TodoStore, {})
  })
  .views(self => ({

  }))
  .actions(self => ({
    afterCreate() {
      //Load the sales data, but delay to simulate HTTP delay
      setTimeout(() => {
        self.chartStore.getSales();
      }, 1000);
    }
  }))