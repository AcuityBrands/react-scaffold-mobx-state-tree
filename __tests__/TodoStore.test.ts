import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree'
import { TodoItem, TodoStore, ITodoItem, ITodoStore } from '../src/models/TodoStore'
import * as fetchMock from 'fetch-mock'

afterEach(function () {
  // Restore the fetch to native implementation after each mock
  afterEach(fetchMock.restore);
})

it("can create an instance of a todo", () => {
  const todo = TodoItem.create({
    id: 1,
    title: "Test Todo",
  });

  expect(todo.title).toBe("Test Todo");
  expect(todo.completed).toBe(false);

  todo.changeLabel("New Label");
  todo.changeComplete(true);

  expect(todo.title).toBe("New Label");
  expect(todo.completed).toBe(true);
})

it("can create an instance of a todo store", () => {
  const store = TodoStore.create({
    todos: [{
      id: 1,
      title: "Test Todo"
    }]
  })

  expect(store.todos.length).toBe(1);
  expect(store.todos[0].title).toBe("Test Todo");
})

// Use async/await to test mocked api call since the mock
// returns a promise
it("can load a set of suggested todos", async () => {
  fetchMock.mock("https://jsonplaceholder.typicode.com/todos", {
    status: 200,
    body: [{
      "userId": 1,
      "id": 1,
      "title": "todo1",
      "completed": false
    }],
  })
  const store = TodoStore.create();
  await store.getSuggestions(1);
  expect(store.todoCount).toBe(1);
  expect(store.todos[0].title).toBe("todo1");
})

it("can add new todos - track snapshots", () => {
  const store = TodoStore.create();
  const states = [];

  onSnapshot(store, snapshot => {
    states.push(snapshot)
  })

  const item = TodoItem.create({
    id: 1,
    title: "Test Todo"
  })
  store.add(item);

  expect(store.todos.length).toBe(1);
  expect(store.todos[0].title).toBe("Test Todo");
  expect(getSnapshot(store)).toEqual({
    loading: false,
    todos: [{
      id: 1,
      userId: 1,
      title: "Test Todo",
      completed: false
    }]
  })
  expect(getSnapshot(store)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
})

it("can add new items - track patches", () => {
  const store = TodoStore.create();
  const patches = [];

  onPatch(store, patch => {
    patches.push(patch)
  })

  store.add({
    id: 1,
    title: "Test Todo",
    completed: false
  } as ITodoItem)

  store.todos[0].changeLabel("New Label");
  expect(patches).toMatchSnapshot();
})

it("can calculate computed values", () => {
  const store = TodoStore.create({
    todos: [{
      id: 1,
      title: "Test Todo",
    }]
  })

  expect(store.todoCount).toBe(1);
  expect(store.completedCount).toBe(0);

  store.todos[0].changeComplete(true);

  expect(store.completedCount).toBe(1);
  expect(store.completedPct).toBe(100);
})