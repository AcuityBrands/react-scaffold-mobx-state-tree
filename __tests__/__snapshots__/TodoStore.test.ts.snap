// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items - track patches 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/todos/0",
    "value": Object {
      "completed": false,
      "id": 1,
      "title": "Test Todo",
      "userId": 1,
    },
  },
  Object {
    "op": "replace",
    "path": "/todos/0/title",
    "value": "New Label",
  },
]
`;

exports[`can add new todos - track snapshots 1`] = `
Object {
  "loading": false,
  "todos": Array [
    Object {
      "completed": false,
      "id": 1,
      "title": "Test Todo",
      "userId": 1,
    },
  ],
}
`;

exports[`can add new todos - track snapshots 2`] = `
Array [
  Object {
    "loading": false,
    "todos": Array [
      Object {
        "completed": false,
        "id": 1,
        "title": "Test Todo",
        "userId": 1,
      },
    ],
  },
]
`;
